torch/_inductor/index_propagation.py

```diff
@@ -267,15 +287,37 @@ def inner(*args: Any, **kwargs: Any) -> IndexPropResult:
 
         return inner
 
+    def statically_true(self, e):
+        evaluated = self.shape_env._maybe_evaluate_static(
+            e,
+            axioms=self.axioms,
+            var_to_range=self.var_to_range,
+        )
+        return bool(evaluated)
+
     def indirect_indexing(
         self, index: Union[Any, IndexPropVar], size: Any, check: bool = True
     ) -> Any:
-        # nb. We do index + Where(...) rather than Where(idx >= 0, idx, idx + sz) because we don't have CSE
-        #     for SymPy expressions, so we don't want to repeat idx too much
-
-        # indirect_indexing returns a sympy value, so no need to wrap in IndexPropVar here
         if isinstance(index, IndexPropVar) and index.is_symbolic:
-            # If we are turning a indirect indexing into direct, we need to wrap it.
-            index = index.value.expr
-            return index + Where(index >= 0, 0, size)
+            expr = sympy.sympify(index.value.expr)
+
+            def wrap_expr(expr):
+                # Positive, negative, mixed
+                if self.statically_true(0 <= expr):
+                    return expr
+                elif self.statically_true(expr < 0):
+                    return expr + size
+                else:
+                    return Where(expr < 0, expr + size, expr)
```

Comment:
Why has this become more complicated? I would expect that `index + Where(index >= 0, 0, size)` would be simplified to `index` if `index >= 0` could be statically proved.

Reply:
Not quite. Sometimes we can prove via the guards that we have in shape_env that something is statically true, but we don't have access to this information within the `Where` class.

If we keep a global `shape_env` that's always updated with the correct variable names at every point in time, it would be possible to subsume this within the `Where` class (and every other sympy class we have), but that would require substantially more effort.

Reply:
Sure but:
1. It would be equivalent to doing it here, as `Where` is just instantiated here
2. We would need to instantiate the relevant shape_env with the axioms there as well.

Given that we already have all the info here, I think it's better to keep it here. When we have a global shape_env for inductor all this can of course be refactored and simplified, but I don't think it's a good idea to start planting this "axioms" pattern all around the place now.

Reply:
>  but I don't think it's a good idea to start planting this "axioms" pattern all around the place now.

Surely if we're going to add it, the `SimplifyIndexing` pass is the place this belongs, no?

Reply:
so should I dupe the shape env + axioms creation and add it to `SimplifyIndexing` as well, you say?